pkg/channel/event_dispatcher.go

```diff
@@ -0,0 +1,165 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package channel
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	cehttp "github.com/cloudevents/sdk-go/pkg/cloudevents/transport/http"
+	"go.opencensus.io/plugin/ochttp/propagation/b3"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/eventing/pkg/utils"
+)
+
+const correlationIDHeaderName = "Knative-Correlation-Id"
+
+type Dispatcher interface {
+	// DispatchEvent dispatches an event to a destination over HTTP.
+	//
+	// The destination and reply are DNS names. For names with a single label,
+	// the default namespace is used to expand it into a fully qualified name
+	// within the cluster.
+	DispatchEvent(ctx context.Context, event cloudevents.Event, destination, reply string, defaults DispatchDefaults) error
+}
+
+// EventDispatcher is the 'real' Dispatcher used everywhere except unit tests.
+var _ Dispatcher = &EventDispatcher{}
+
+var propagation = &b3.HTTPFormat{}
+
+// EventDispatcher dispatches events to a destination over HTTP.
+type EventDispatcher struct {
+	ceClient         cloudevents.Client
+	supportedSchemes sets.String
+
+	logger *zap.Logger
+}
+
+// DispatchDefaults provides default parameter values used when dispatching an event.
+type DispatchDefaults struct {
+	Namespace string
+}
+
+// NewEventDispatcher creates a new event dispatcher that can dispatch
+// events to HTTP destinations.
+func NewEventDispatcher(logger *zap.Logger) *EventDispatcher {
+	ceClient, err := kncloudevents.NewDefaultClient()
+	if err != nil {
+		logger.Fatal("failed to create cloudevents client", zap.Error(err))
+	}
+	return &EventDispatcher{
+		ceClient:         ceClient,
+		supportedSchemes: sets.NewString("http", "https"),
+		logger:           logger,
+	}
+}
+
+// DispatchEvent dispatches an event to a destination over HTTP.
+//
+// The destination and reply are DNS names. For names with a single label,
+// the default namespace is used to expand it into a fully qualified name
+// within the cluster.
+func (d *EventDispatcher) DispatchEvent(ctx context.Context, event cloudevents.Event, destination, reply string, defaults DispatchDefaults) error {
+	var err error
+	// Default to replying with the original event. If there is a destination, then replace it
+	// with the response from the call to the destination instead.
+	response := &event
+	if destination != "" {
+		destinationURL := d.resolveURL(destination, defaults.Namespace)
+		ctx, response, err = d.executeRequest(ctx, destinationURL, event)
+		if err != nil {
+			return fmt.Errorf("unable to complete request to %s: %v", destinationURL, err)
+		}
+	}
+
+	if reply != "" && response != nil {
```

Comment:
I would log the response when it is not nil and reply is nil

Reply:
Done, thanks!